# SJE/ex01

https://github.com/takatakamanbou/SJE

## 準備

以下のリンク先の画像をダウンロードして，この notebook と同じフォルダ/ディレクトリに置きましょう．
リンクをたどって画像が表示されたら，画像を右クリックして「名前を付けて画像を保存」とかそんなん．

https://github.com/takatakamanbou/SJE/blob/main/blackuni3.png



この notebook では，**OpenCV** と **NumPy** を使います．

- OpenCV:  コンピュータビジョンのためのライブラリ．C++ や Python から呼び出して画像処理などいろんなことができる．
- NumPy: Python 用の科学技術計算パッケージ．

以下のセルは，そのための準備です．実行してから先へ進みましょう．

In [1]:
import cv2  # OpenCV のパッケージをインポート
import numpy as np  # NumPy のパッケージを np という名前でインポート

## 課題A
以下の内容を理解しながら順次実行していこう．

以下を実行すると，`blackuni3.png` という画像ファイルが読み込まれます．この画像の画素値が `img` という名前の変数に代入されます．

In [2]:
img = cv2.imread('blackuni3.png')

この画像は，幅170画素，高さ120画素のカラー画像です，変数 `img` は三次元の配列（より詳しく言うと NumPy の array というデータ構造）になっています．次のセルを実行すると，中身を表示させることができます．

In [3]:
print(img)

[[[166 160 161]
  [164 158 159]
  [161 155 156]
  ...
  [185 167 166]
  [170 152 151]
  [161 143 142]]

 [[174 168 169]
  [161 155 156]
  [152 146 147]
  ...
  [173 155 154]
  [169 151 150]
  [161 143 142]]

 [[162 156 157]
  [156 150 151]
  [150 144 145]
  ...
  [151 133 132]
  [167 149 148]
  [166 148 147]]

 ...

 [[164 154 154]
  [165 155 155]
  [154 144 144]
  ...
  [128 120 121]
  [124 116 117]
  [138 130 131]]

 [[164 156 156]
  [165 156 156]
  [161 153 153]
  ...
  [132 126 127]
  [120 114 115]
  [133 127 128]]

 [[160 152 152]
  [177 169 169]
  [178 170 170]
  ...
  [136 130 131]
  [130 124 125]
  [129 123 124]]]


いちばん最初の `[166 160 161]` は，この画像の左上の画素の値が，青(B) = 166, 緑(G) = 160, 赤(R) = 161 であることを表しています．
次の `[164 158 159]` は，その右隣の画素の値．
以下，左上から右に向かう順に画素値が並んでいます．

次のセルを実行すると，配列 `img` の形(shape)がわかります．

In [4]:
print(img.shape)

(120, 170, 3)


配列（NumPy array）に対して，次のように `[ ]` を使うと，配列の中の個々の要素の値を得ることができます．

In [5]:
print(img[0, 0, 0]) # 左上隅の画素の B の値
print(img[0, 1, 2]) # その右隣の画素の R の値

166
159


また，次のようにするとBGRの値をまとめて得ることもできます，

In [6]:
print(img[0, 0]) # 左上隅
print(img[0, 169]) # 幅170画素の画像なので，これは右上隅

[166 160 161]
[161 143 142]


以下のセルの中身を編集して，左下隅と右下隅の画素値を表示させましょう．

In [7]:
print(img[0, 0]) # 左下隅
print(img[0, 0]) # 右下隅

[166 160 161]
[166 160 161]


OpenCV には様々な関数があります．たとえば，`cv2.flip` という関数を使うと，画像を左右（や上下に）反転させることができます．
以下のセルを実行すると `hoge.png` というファイルができるはずですので，表示させてみましょう．

In [8]:
img2 = cv2.flip(img, 1) # img の左右を反転したものを img2 とする
cv2.imwrite('hoge.png', img2) # img2 を `hoge.png` というファイル名で保存

True

`cv2.flip` の第二引数の値を変えると，上下に反転させたりもできます．どういう値を指定すればよいか調べて実際にやってみよう．

配列 `img` の要素は，0 から 255 までの整数値です．RGBそれぞれの値を256段階で表しているわけです．`[0, 0, 0]` は黒で，`[255, 255, 255]` が白です．以下のセルを実行すると，すべての画素値が

`(新しい画素値) ← 255 - （元の画素値）`

となります．何が起こるかやってみましょう．

In [9]:
img2 = 255 - img
cv2.imwrite('hoge.png', img2)

True

NumPy の array では，**スライス** という機能が使えます．

In [10]:
a = np.arange(15) # 0 からはじまる15個の整数値をならべた Numpy array
print(a)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14]


In [11]:
print(a[3:6]) # 3番目の要素から 5番目（6の一つ手前）の要素まで
print(a[:6]) # 最初から5番目の要素まで
print(a[8:]) # 8番目から最後の要素まで

[3 4 5]
[0 1 2 3 4 5]
[ 8  9 10 11 12 13 14]


In [12]:
A = a.reshape(3, 5) # 配列 a を 3 行 5 列にならべかえた二次元の Numpy array
print(A)

[[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]]


In [13]:
print(A[1:, 1:3]) # 1行目以降，1列目から2列目まで

[[ 6  7]
 [11 12]]


In [14]:
print(A[:, 0]) # 0列目ぜんぶ

[ 0  5 10]


スライスで指定した範囲に値を代入することもできます．

In [15]:
A[1:, 1:3] = 99
print(A)

[[ 0  1  2  3  4]
 [ 5 99 99  8  9]
 [10 99 99 13 14]]


In [16]:
A[:, 0] = [0, 111, 222] # 0列目に順に 0, 111, 222 を代入
print(A)

[[  0   1   2   3   4]
 [111  99  99   8   9]
 [222  99  99  13  14]]


上記のように `A` をいじったあとで `a` の中身を確認すると...．詳しい説明は省きますが，ここでの変数 `A` は 変数 `a` とメモリ領域を共有しているのでこういうことになります．

In [17]:
print(a)

[  0   1   2   3   4 111  99  99   8   9 222  99  99  13  14]


上記の技を使って，画像をちょろっといじってみましょう．

In [18]:
img = cv2.imread('blackuni3.png')
img[30:40, 80:120, :] = [255, 0, 255] # 画像の一部をマゼンタ（赤紫）で塗りつぶす
cv2.imwrite('hoge.png', img)

True

## 課題B